src/bdecode.cpp

```diff
@@ -289,6 +289,93 @@ namespace {
 		m_buffer = buf;
 	}
 
+	bool bdecode_node::has_soft_error(span<char> error) const
+	{
+		bdecode_token const* tokens = m_root_tokens;
+		std::uint32_t token = std::uint32_t(m_token_idx);
+
+		// we don't know what the original depth_limit was
+		// so this has to go on the heap
+		std::vector<uint32_t> stack;
+		// make the initial allocation the default depth_limit
+		stack.reserve(100);
+
+		do
+		{
+			switch (tokens[token].type)
+			{
+			case bdecode_token::integer:
+				if (m_buffer[tokens[token].offset + 1] == '0'
+					&& m_buffer[tokens[token].offset + 2] != 'e')
+				{
+					std::snprintf(error.data(), error.size(), "leading zero in integer");
+					return true;
+				}
+				break;
+			case bdecode_token::string:
+				if (m_buffer[tokens[token].offset] == '0'
+					&& m_buffer[tokens[token].offset + 1] != ':')
+				{
+					std::snprintf(error.data(), error.size(), "leading zero in string length");
+					return true;
+				}
+				break;
+			case bdecode_token::dict:
+			case bdecode_token::list:
+				stack.push_back(token);
+				break;
+			case bdecode_token::end:
+				auto const parent = stack.back();
+				stack.pop_back();
+				if (tokens[parent].type == bdecode_token::dict
+					&& token != parent + 1)
+				{
+					// this is the end of a non-empty dict
+					// check the sort order of the keys
+					std::uint32_t k1 = parent + 1;
+					for (;;)
+					{
+						// skip to the first key's value
+						std::uint32_t const v1 = k1 + tokens[k1].next_item;
+						// then to the next key
+						std::uint32_t const k2 = v1 + tokens[v1].next_item;
+
+						// check if k1 was the last key in the dict
+						if (k2 == token)
+							break;
+
+						std::uint32_t const v2 = k2 + tokens[k2].next_item;
+
+						std::uint32_t const k1_start = tokens[k1].offset + std::uint32_t(tokens[k1].start_offset());
+						std::uint32_t const k1_len = tokens[v1].offset - k1_start;
+						std::uint32_t const k2_start = tokens[k2].offset + std::uint32_t(tokens[k2].start_offset());
+						std::uint32_t const k2_len = tokens[v2].offset - k2_start;
+
+						std::uint32_t const min_len = std::min(k1_len, k2_len);
+
+						int cmp = std::memcmp(m_buffer + k1_start, m_buffer + k2_start, min_len);
```

Comment:
oh, I see, this is enforcing the lexicographical ordering of keys, so you actually compare greater than